### PR TITLE
Cleanups and enhancements for `basic_string`

### PR DESCRIPTION
**Pull request by achabense**
- Merge `_Memcpy_val_from` to its only caller.
- Simplify `basic_string(initializer_list)`
- Enhance exception guarantee for `basic_string(&&, off)`. The rvalue will be unchanged if exception is thrown from `_Alloc_proxy()`.
- `noexcept` enhancements, mainly for comparisons with c-string.
- Some other nitpicks.

### CODE REVIEW

**Comment by achabense** [comment]
https://github.com/microsoft/STL/blob/60f18856c56389001c38a9929ef37bd5c01c4e47/stl/inc/xstring#L2785-L2786
I think this `noexcept` enhancement is not suitable, as it is not mandated by the standard, and involves `_Alloc_proxy`.

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:4279]: https://github.com/microsoft/STL/blob/60f18856c56389001c38a9929ef37bd5c01c4e47/stl/inc/xstring#L4272
Is it encouraged to move this `using _STD swap` into the if block where it is used? I'm not sure as find they are usually defined at the beginning of functions.

File: stl/inc/xstring
Comment (StephanTLavavej) [reply stl/inc/xstring:2733]: I think `+ 1` makes sense here.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
All of these cleanups look great - thanks!

**Comment by StephanTLavavej** [comment]
I have enough confidence in these cleanups to merge them without a second maintainer approval.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
:thread: :yarn: :broom:
===

**Pull request merged**
